### PR TITLE
Support e2e run configs in linux benchmark tool

### PR DESCRIPTION
**Pull request by pzread**
Add `--run_config` option to take `E2EModelRunConfig` in linux benchmark tool.

### CODE REVIEW

File: build_tools/benchmarks/common/common_arguments.py [hunk 1]
```diff
@@ -129,6 +136,10 @@ def check_exe_path(path):
       "for). In that case, no --benchmark_repetitions flag will be passed."
       " If not specified, a --benchmark_repetitions will be passed "
       "instead.")
+  parser.add_argument("--run_config",
+                      type=check_file_path,
+                      default=None,
+                      help="JSON file of the run configs")
```
Comment (GMNGeoffrey) [inline build_tools/benchmarks/common/common_arguments.py:142]: I'm confused by the plural here. Is it one run configuration or multiple??

File: build_tools/benchmarks/run_benchmarks_on_linux.py [hunk 2]
```diff
@@ -12,19 +12,24 @@
...
-import subprocess
+from typing import Any, List, Optional
 import atexit
+import json
 import shutil
+import subprocess
 import tarfile
-
-from typing import List, Optional
+import typing
```
Comment (GMNGeoffrey) [inline build_tools/benchmarks/run_benchmarks_on_linux.py:21]: I think Google style would put the `typing` imports together: https://google.github.io/styleguide/pyguide.html#313-imports-formatting. I thought it called for direct symbol imports in their own section, but it looks like I made that up.

File: build_tools/benchmarks/run_benchmarks_on_linux.py [hunk 3]
```diff
@@ -38,48 +43,69 @@ def run_benchmark_case(self, benchmark_case: BenchmarkCase,
...
-                         run_flags=run_flags)
+                         capture_filename=capture_filename)
 
   def __parse_flagfile(self, case_dir: pathlib.Path) -> List[str]:
     return [
         line.strip()
         for line in (case_dir / MODEL_FLAGFILE_NAME).read_text().splitlines()
     ]
 
+  def __build_tool_cmds(self, benchmark_case: BenchmarkCase,
```
Comment (GMNGeoffrey) [inline build_tools/benchmarks/run_benchmarks_on_linux.py:60]: Double underscore dunders are discouraged by Google style: https://google.github.io/styleguide/pyguide.html#3162-naming-conventions. Looks like you already have some of these, so best leave it for now for consistency, but may want to switch to single underscore in a follow-up if there's not some strong reason to dunder here

File: build_tools/benchmarks/run_benchmarks_on_linux.py [hunk 4]
```diff
@@ -38,48 +43,69 @@ def run_benchmark_case(self, benchmark_case: BenchmarkCase,
...
+    run_config = benchmark_case.run_config
+    if run_config is None:
+      # TODO(#11076): Remove legacy path.
+      if benchmark_case.benchmark_case_dir is None:
+        raise ValueError(
+            "benchmark_case_dir can't be None if run_config is None.")
+
+      # TODO(pzread): Taskset should be derived from CPU topology.
+      # Only use the low 8 cores.
+      cmds = ["taskset", "0xFF", tool_path]
```
Comment (GMNGeoffrey) [inline build_tools/benchmarks/run_benchmarks_on_linux.py:71]: Shouldn't this be part of the run configuration? So we're not going to later derive it from CPU topology: we're just going to delete this entire if block, right?

File: build_tools/benchmarks/run_benchmarks_on_linux.py [hunk 5]
```diff
@@ -38,48 +43,69 @@ def run_benchmark_case(self, benchmark_case: BenchmarkCase,
...
+
+      run_flags = self.__parse_flagfile(benchmark_case.benchmark_case_dir)
+      # Replace the CUDA device flag with the specified GPU.
+      if benchmark_case.driver_info.driver_name == "cuda":
+        run_flags = list(
+            filter(lambda flag: not flag.startswith("--device"), run_flags))
```
Comment (GMNGeoffrey) [inline build_tools/benchmarks/run_benchmarks_on_linux.py:77]: I think list comprehensions are considered more "Pythonic"

```suggestion
        run_flags = [flag for flag in run_flags if not flag.startswith("--device")]
```

File: build_tools/benchmarks/run_benchmarks_on_linux.py [hunk 6]
```diff
@@ -38,48 +43,69 @@ def run_benchmark_case(self, benchmark_case: BenchmarkCase,
...
+        run_flags.append(f"--device=cuda://{self.gpu_id}")
+
+      return cmds + run_flags
+
+    cmds: List[Any] = run_module_utils.build_linux_wrapper_cmds_for_device_spec(
+        run_config.target_device_spec)
+    cmds += [tool_path]
```
Comment (GMNGeoffrey) [inline build_tools/benchmarks/run_benchmarks_on_linux.py:84]: append?

```suggestion
    cmds.append(tool_path)
```

(same below)

File: build_tools/python/e2e_model_tests/run_module_utils.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    module_execution_config: ModuleExecutionConfig,
+    gpu_id: str = "0") -> List[str]:
+  """Returns the IREE run module flags of the execution config."""
+
+  run_flags = list(module_execution_config.extra_flags)
+  driver = module_execution_config.driver
+  if driver == RuntimeDriver.CUDA:
+    run_flags.append(f"--device=cuda://{gpu_id}")
+  else:
+    run_flags.append(f"--device={driver.value}")
```
Comment (GMNGeoffrey) [inline build_tools/python/e2e_model_tests/run_module_utils.py:39]: Hmmm this is ok for now, but feels a bit hacky. Would be nice to figure out a cleaner way to encode this information in the right place.

File: build_tools/python/e2e_model_tests/run_module_utils_test.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import unittest
+
+from e2e_model_tests import run_module_utils
+from e2e_test_framework.definitions import common_definitions, iree_definitions
+from e2e_test_framework.device_specs import device_parameters
+
+
+class RunModuleTuilsTest(unittest.TestCase):
```
Comment (GMNGeoffrey) [inline build_tools/python/e2e_model_tests/run_module_utils_test.py:14]: Typo?

```suggestion
class RunModuleUtilsTest(unittest.TestCase):
```

File: build_tools/python/e2e_model_tests/run_module_utils_test.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        tags=[],
+        source_type=common_definitions.ModelSourceType.EXPORTED_TFLITE,
+        source_url="https://example.com/xyz.tflite",
+        entry_function="main",
+        input_types=["1xf32", "2x2xf32"])
+
+    flags = run_module_utils.build_run_flags_for_model(
+        model, common_definitions.ZEROS_MODEL_INPUT_DATA)
+
+    self.assertEqual(flags, [
```
Comment (GMNGeoffrey) [inline build_tools/python/e2e_model_tests/run_module_utils_test.py:29]: Does order matter? If not, use `assertCountEqual` (name is confusing IMO, but it's say the two sequences have equal counts of each item)

**Review by GMNGeoffrey — APPROVED** [review]

File: build_tools/benchmarks/run_benchmarks_on_linux.py (hunk 4 above)
Comment (pzread) [reply build_tools/benchmarks/run_benchmarks_on_linux.py:71]: This is the legacy path, will be removed during the cleanup.

File: build_tools/benchmarks/run_benchmarks_on_linux.py (hunk 2 above)
Comment (pzread) [reply build_tools/benchmarks/run_benchmarks_on_linux.py]: Done.

File: build_tools/benchmarks/run_benchmarks_on_linux.py (hunk 5 above)
Comment (pzread) [reply build_tools/benchmarks/run_benchmarks_on_linux.py]: Done.

File: build_tools/python/e2e_model_tests/run_module_utils_test.py (hunk 9 above)
Comment (pzread) [reply build_tools/python/e2e_model_tests/run_module_utils_test.py:29]: In some cases the order does matter (e.g. multiple `--function_input=` should be in order). I'll keep the `assertEqual` to make sure any change on the order is verified.

**Pull request merged**